File: app/src/main/java/com/rgapro1/ocaso/PolicyProductSchema.java
package com.rgapro1.ocaso;

import java.util.*;

/** Canonical schema: classify product before persisting OCR fields. */
public final class PolicyProductSchema {
    private PolicyProductSchema() {}

    public enum ProductType { VIDA, DECESOS, AHORRO, HOGAR, ACCIDENTES, DESCONOCIDO }

    public static final class Common {
        public String policyNumber;
        public String holderName;
        public String holderDni;
        public String issueDate;
        public String effectiveDate;
        public String expiryDate;
        public String renewalDate;
    }

    public static final class Insured {
        public String name;
        public String birthDate;
        public String dni;
        public String capital;
    }

    public static ProductType classify(String text) {
        String t = text == null ? "" : text.toLowerCase(Locale.ROOT);
        if (containsAny(t, "decesos", "asistencia familiar", "servicio funerario")) return ProductType.DECESOS;
        if (containsAny(t, "seguro de vida", "vida riesgo", "capital por fallecimiento")) return ProductType.VIDA;
        if (containsAny(t, "ahorro", "aportación periódica", "aportacion periodica")) return ProductType.AHORRO;
        if (containsAny(t, "hogar", "continente", "contenido")) return ProductType.HOGAR;
        if (containsAny(t, "accidentes", "accidente personal")) return ProductType.ACCIDENTES;
        return ProductType.DESCONOCIDO;
    }

    private static boolean containsAny(String text, String... values) {
        for (String value : values) if (text.contains(value)) return true;
        return false;
    }

    /** Only the common fields belong to every product. Product-specific parsers add their own data. */
    public static Set<String> allowedCommonFields() {
        return new LinkedHashSet<>(Arrays.asList(
                "policyNumber", "holderName", "holderDni", "issueDate",
                "effectiveDate", "expiryDate", "renewalDate"));
    }
}

File: app/src/main/java/com/example/rgapro1/PolicyProductSchema.kt
package com.example.rgapro1

/**
 * Canonical product-specific extraction schema.
 * OCR text must be classified before fields are persisted.
 */
enum class PolicyProductType {
    VIDA,
    DECESOS,
    AHORRO,
    HOGAR,
    ACCIDENTES,
    DESCONOCIDO
}

data class PolicyCommonFields(
    val policyNumber: String? = null,
    val holderName: String? = null,
    val holderDni: String? = null,
    val issueDate: String? = null,
    val effectiveDate: String? = null,
    val expiryDate: String? = null,
    val renewalDate: String? = null
)

data class LifePolicyFields(
    val holderAddress: String? = null,
    val holderPhone: String? = null,
    val holderEmail: String? = null,
    val insuredCapitals: Map<String, String> = emptyMap()
)

data class FuneralInsured(
    val name: String,
    val birthDate: String? = null,
    val dni: String? = null,
    val capital: String? = null
)

data class FuneralPolicyFields(
    val insured: List<FuneralInsured> = emptyList()
)

data class SavingsPolicyFields(
    val contribution: String? = null,
    val guarantees: Map<String, String> = emptyMap()
)

data class StructuredPolicy(
    val productType: PolicyProductType,
    val common: PolicyCommonFields = PolicyCommonFields(),
    val life: LifePolicyFields? = null,
    val funeral: FuneralPolicyFields? = null,
    val savings: SavingsPolicyFields? = null,
    val rawOcrText: String? = null,
    val originalDocumentUri: String? = null
)

object PolicyProductClassifier {
    fun classify(text: String): PolicyProductType {
        val normalized = text.lowercase()
        return when {
            listOf("decesos", "funeral", "asistencia familiar").any(normalized::contains) -> PolicyProductType.DECESOS
            listOf("seguro de vida", "vida riesgo", "fallecimiento").any(normalized::contains) -> PolicyProductType.VIDA
            listOf("ahorro", "aportación", "aportacion", "prima periódica", "prima periodica").any(normalized::contains) -> PolicyProductType.AHORRO
            listOf("hogar", "vivienda", "continente", "contenido").any(normalized::contains) -> PolicyProductType.HOGAR
            listOf("accidentes", "accidente personal").any(normalized::contains) -> PolicyProductType.ACCIDENTES
            else -> PolicyProductType.DESCONOCIDO
        }
    }
}
